engine/remotecache/cache.go

```diff
@@ -83,6 +95,108 @@ func ResolveCacheImporterFunc(sm *session.Manager, cs content.Store, hosts docke
 	}
 }
 
+func s3Exporter(ctx context.Context, g session.Group, attrs map[string]string) (remotecache.Exporter, error) {
+	// For now, if manifests_prefix is set but name isn't, export to a random name to support pooling of manifests under that prefix
+	// TODO: make a less bad interface
+	if attrs["manifests_prefix"] != "" && attrs["name"] == "" {
+		id := identity.NewID()
+		attrs["name"] = id
+	}
+	return s3remotecache.ResolveCacheExporterFunc()(ctx, g, attrs)
+}
+
+func s3Importer(ctx context.Context, g session.Group, attrs map[string]string) (remotecache.Importer, ocispecs.Descriptor, error) {
+	// For now, if manifests_prefix is set but name isn't, import everything from under that prefix
+	if attrs["manifests_prefix"] != "" && attrs["name"] == "" {
+		bklog.G(ctx).Debugf("importing all manifests under prefix %s", attrs["manifests_prefix"])
+		region := attrs["region"]
+		if region == "" {
+			region = os.Getenv("AWS_REGION")
+		}
+		bucket := attrs["bucket"]
+		if bucket == "" {
+			bucket = os.Getenv("AWS_BUCKET")
+		}
+
+		globalPrefix := attrs["prefix"]
+		manifestsPrefix := attrs["manifests_prefix"]
+
+		accessKey := attrs["access_key_id"]
+		secretKey := attrs["secret_access_key"]
+		sessionToken := attrs["session_token"]
+		endpointURL := attrs["endpoint_url"]
+		usePathStyle := attrs["use_path_style"] == "true"
+
+		cfg, err := awsConfig.LoadDefaultConfig(ctx, awsConfig.WithRegion(region))
+		if err != nil {
+			return nil, ocispecs.Descriptor{}, errors.Errorf("Unable to load AWS SDK config, %v", err)
+		}
+		client := s3.NewFromConfig(cfg, func(options *s3.Options) {
+			if accessKey != "" && secretKey != "" {
+				options.Credentials = credentials.NewStaticCredentialsProvider(accessKey, secretKey, sessionToken)
+			}
+			if endpointURL != "" {
+				options.UsePathStyle = usePathStyle
+				options.EndpointResolver = s3.EndpointResolverFromURL(endpointURL)
+			}
+		})
+
+		prefix := globalPrefix + manifestsPrefix
+		listResp, err := client.ListObjectsV2(ctx, &s3.ListObjectsV2Input{
```

Comment:
Oh yeah none of this code scales to that level right now, we'll honestly run into more issues before we hit 1000 manifests to import I think. But the aws sdk has builtin pagination helpers that are trivial to setup so I updated with that for now anyways.

I think your idea of hashing the manifest and using that as the name can help stave off these problems for a bit longer. I'll probably save that for follow ups though since it'll require copying in more code from buildkit.

After that yeah we'll need to decide how to either gc and/or merge manifests so the pool doesn't grow forever. Many options between having clients do it (while not stomping over each other), configuring s3 object expiration, a centralized "janitor" that users run, or even going all the way to just managing these manifests via a saas.